### PR TITLE
feat: progressive billing (plan + subscription)

### PR DESCRIPTION
**Pull request by keellyp**
## Roadmap Task

LAGO-222

## Context

Progressive billing feature

## Description

Add the possibility to add progressive billing in plan

Note: I recommend reviewing by files changed instead of commit by commit as I've refactored things during development

### CODE REVIEW

File: src/components/invoices/details/InvoiceDetailsTableFooter.tsx [hunk 1]
```diff
@@ -51,6 +52,32 @@ export const InvoiceDetailsTableFooter = memo(
...
+              <tr>
+                <td></td>
+                <td colSpan={colSpan}>
+                  <Typography variant="bodyHl" color="grey600">
+                    {translate('text_1724936123802q74m2xxak3o')}
+                  </Typography>
+                </td>
+                <td>
+                  <Typography variant="body" color="success600">
+                    {' '}
```
Comment (ansmonjol) [inline src/components/invoices/details/InvoiceDetailsTableFooter.tsx:65]: Is that intended?

File: src/components/plans/details/PlanDetailsAdvancedSettingsSection.tsx [hunk 2]
```diff
@@ -79,8 +79,7 @@ const PlanDetailsAdvancedSettingsSection = ({
                             currency: currency,
                           },
                         ),
-                        threshold.thresholdDisplayName ||
-                          `${translate('text_1724179887723pt34ivcecdt')} ${i + 1}`,
+                        threshold.thresholdDisplayName || '',
```
Comment (ansmonjol) [inline src/components/plans/details/PlanDetailsAdvancedSettingsSection.tsx:82]: I thought we should display a placeholder name here, did that changed in the meantime?

File: src/core/serializers/serializePlanInput.ts [hunk 3]
```diff
@@ -144,6 +146,24 @@ export const serializePlanInput = (values: PlanFormInput) => {
             taxes: undefined,
           }
         : {},
+    usageThresholds: [
+      ...(nonRecurringUsageThresholds ?? []).map((threshold) => ({
+        ...threshold,
+        thresholdDisplayName: threshold.thresholdDisplayName ?? null,
```
Comment (ansmonjol) [inline src/core/serializers/serializePlanInput.ts:152]: hum, If the user tries to edit the name, and then remove the value (at the same moment), the value will be an empty string if I'm correct
And there you will then send this empty string to the BE.
How would that be then on fetch? Will we receive `null` or `''`?
If it's null, maybe we should use `||` instead WDYT?

File: src/components/plans/details/PlanDetailsAdvancedSettingsSection.tsx (hunk 2 above)
Comment (keellyp) [reply src/components/plans/details/PlanDetailsAdvancedSettingsSection.tsx:82]: Ye it did (cf. https://linear.app/getlago/issue/LAGO-256/%5Bqa-8%5D-remove-predefined-name-when-threshold-name-is-empty)

File: src/core/serializers/serializePlanInput.ts [hunk 4]
```diff
@@ -144,6 +146,24 @@ export const serializePlanInput = (values: PlanFormInput) => {
...
+        amountCents: Number(serializeAmount(threshold.amountCents, values.amountCurrency)),
+      })),
+      ...(recurringUsageThreshold
+        ? [
+            {
+              ...recurringUsageThreshold,
+              thresholdDisplayName: recurringUsageThreshold.thresholdDisplayName ?? null,
```
Comment (ansmonjol) [inline src/core/serializers/serializePlanInput.ts:159]: same as above here

File: src/components/invoices/details/InvoiceDetailsTableFooter.tsx (hunk 1 above)
Comment (keellyp) [reply src/components/invoices/details/InvoiceDetailsTableFooter.tsx:65]: Nop, fixing it!

File: src/components/plans/details/PlanDetailsAdvancedSettingsSection.tsx (hunk 2 above)
Comment (ansmonjol) [reply src/components/plans/details/PlanDetailsAdvancedSettingsSection.tsx:82]: ok!

**Review by ansmonjol — APPROVED** [review]

**Pull request merged**